# 多元线性回归/对数线性回归（二选一）

## 一、多元线性回归
这部分的内容是要求大家完成多元线性回归，我们会先带着大家使用sklearn做一元线性回归的十折交叉验证，多元线性回归大家可以仿照着完成

### 1. 读取数据

In [15]:
import numpy as np

In [16]:
import pandas as pd

# 读取数据
data = pd.read_csv('data/kaggle_house_price_prediction/kaggle_hourse_price_train.csv')

# 丢弃有缺失值的特征（列）
data.dropna(axis = 1, inplace = True)

# 只保留整数的特征
data = data[[col for col in data.dtypes.index if data.dtypes[col] == 'int64']]

In [17]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 35 columns):
Id               1460 non-null int64
MSSubClass       1460 non-null int64
LotArea          1460 non-null int64
OverallQual      1460 non-null int64
OverallCond      1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
BsmtFinSF1       1460 non-null int64
BsmtFinSF2       1460 non-null int64
BsmtUnfSF        1460 non-null int64
TotalBsmtSF      1460 non-null int64
1stFlrSF         1460 non-null int64
2ndFlrSF         1460 non-null int64
LowQualFinSF     1460 non-null int64
GrLivArea        1460 non-null int64
BsmtFullBath     1460 non-null int64
BsmtHalfBath     1460 non-null int64
FullBath         1460 non-null int64
HalfBath         1460 non-null int64
BedroomAbvGr     1460 non-null int64
KitchenAbvGr     1460 non-null int64
TotRmsAbvGrd     1460 non-null int64
Fireplaces       1460 non-null int64
GarageCars       1460 non-null int64
Garag

### 2. 引入模型

In [18]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_predict

### 3. 使用sklearn完成一元线性回归的十折交叉验证验证

#### 创建模型

In [19]:
model = LinearRegression()

#### 选取数据

In [20]:
features = ['LotArea']
x = data[features]
y = data['SalePrice']

#### 做十折交叉验证的预测

In [21]:
prediction = cross_val_predict(model, x, y, cv = 10)

这十折交叉验证是按顺序做的，会先将前10%的数据作为测试集，然后会往后顺延到10%到20%，最后将这十份的预测值按顺序拼接后返回

In [22]:
prediction.shape

(1460,)

### 4. 计算评价指标

#### MAE

In [23]:
mean_absolute_error(prediction, data['SalePrice'])

55394.44195244894

#### RMSE

In [24]:
mean_squared_error(prediction, data['SalePrice']) ** 0.5

77868.51337752416

### 5. 请你选择多种特征进行组合，完成多元线性回归，并对比不同的特征组合，它们训练出的模型在十折交叉验证上MAE与RMSE的差别，至少完成3组

###### 扩展：多项式回归（一元线性回归的扩展），尝试对部分特征进行变换，如将其二次幂，三次幂作为特征输入模型，观察模型在预测能力上的变化
###### 提示：多元线性回归，只要在上方的features这个list中，加入其他特征的名字就可以

In [25]:
# YOUR CODE HERE
feature = ["LotArea","OverallQual","EnclosedPorch"]
y = data["SalePrice"]
print(np.array(x))
prediction = cross_val_predict(model,x,y,cv=10)
print("model1:")
print(mean_absolute_error(prediction,data["SalePrice"]))
print(mean_squared_error(prediction,data["SalePrice"])**0.5)
print("model2:")
feature = ["LotArea","OverallQual","EnclosedPorch","YearBuilt","EnclosedPorch"]
x = data[feature]
y = data["SalePrice"]
prediction = cross_val_predict(model,x,y,cv=10)
print(mean_absolute_error(prediction,data["SalePrice"]))
print(mean_squared_error(prediction,data["SalePrice"])**0.5)
print("model3:")
feature = ["LotArea","OverallQual","EnclosedPorch","YearBuilt","EnclosedPorch","YearRemodAdd","KitchenAbvGr"]
x = data[feature]
y = data["SalePrice"]
prediction = cross_val_predict(model,x,y,cv=10)
print(mean_absolute_error(prediction,data["SalePrice"]))
print(mean_squared_error(prediction,data["SalePrice"])**0.5)
print("model4:")
feature = ["LotArea","OverallQual","EnclosedPorch","YearBuilt","EnclosedPorch","YearRemodAdd","KitchenAbvGr"]
x = data[feature]
y = data["SalePrice"]
x["OverallQual"] = x["OverallQual"].map(lambda x: x**2)
x["YearRemodAdd"] = x["YearRemodAdd"].map(lambda x: x**3)
prediction = cross_val_predict(model,x,y,cv=10)
print(mean_absolute_error(prediction,data["SalePrice"]))
print(mean_squared_error(prediction,data["SalePrice"])**0.5)

[[ 8450]
 [ 9600]
 [11250]
 ...
 [ 9042]
 [ 9717]
 [ 9937]]
model1:
55394.44195244894
77868.51337752416
model2:
31338.157784701623
46602.37045006612
model3:
31033.917773097157
46312.98615525729
model4:
28991.550398265495
43589.2691178213


c:\users\tht\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:27: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
c:\users\tht\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:28: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


###### 双击此处填写
1. 模型1使用的特征："LotArea","OverallQual","EnclosedPorch"
2. 模型2使用的特征："LotArea","OverallQual","EnclosedPorch","YearBuilt","EnclosedPorch"
3. 模型3使用的特征："LotArea","OverallQual","EnclosedPorch","YearBuilt","EnclosedPorch","YearRemodAdd","KitchenAbvGr"
4. 模型4使用的特征："LotArea","OverallQual","EnclosedPorch","YearBuilt","EnclosedPorch","YearRemodAdd","KitchenAbvGr"
**其中模型四的"OverallQual"输入值为其二次幂，"YearRemodAdd"为其三次幂**

模型|MAE|RMSE
-|-|-
模型1 | 31936.777534526547 | 47040.736137516506
模型2 | 31338.157784701623 | 46602.37045006612
模型3 | 31033.917773097157 | 46312.98615525729
模型4 | 28991.550398265495 | 43589.2691178213